substrate/network-libp2p/src/timeouts.rs

```diff
@@ -0,0 +1,88 @@
+// Copyright 2018 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.?
+
+use futures::{future, Future, stream, Stream};
+use std::io::Error as IoError;
+use std::time::Instant;
+use tokio_core::reactor::{Handle, Timeout};
+
+/// Builds the timeouts system.
+///
+/// The `timeouts_rx` should be a stream receiving newly-created timeout
+/// requests. Returns a stream that produces items as their timeout elapses.
+/// `T` can be anything you want, as it is transparently passed from the input
+/// to the output.
+pub fn build_timeouts_stream<T>(
+	core: Handle,
+	timeouts_rx: impl Stream<Item = (Instant, T), Error = ()>
+) -> impl Stream<Item = T, Error = IoError> {
+	let next_timeout = next_in_timeouts_stream(timeouts_rx);
```

Comment:
`next_in_timeouts_stream` will panic if `timeouts_rx` throws an error. But the type of `timeouts_rx` isn't known. Can you prove this assumption?